File: app/pipeline_trace.py
"""Append-only text log for end-to-end pipeline latency (Python + Unity)."""

from __future__ import annotations

import datetime
import os
import threading


class PipelineTrace:
    """Thread-safe timestamped trace file (one line per event)."""

    def __init__(self, path: str | None):
        self._path = (path or "").strip()
        self._lock = threading.Lock()
        self._file = None
        if not self._path:
            return
        parent = os.path.dirname(os.path.abspath(self._path))
        if parent:
            os.makedirs(parent, exist_ok=True)
        self._file = open(self._path, "a", encoding="utf-8", buffering=1)

    @property
    def enabled(self) -> bool:
        return self._file is not None

    def log_session_start(self, **metadata: object) -> None:
        """Log run boundary with command line and settings (call once per pose.py launch)."""
        self.log("session_start", seq=-1, source="python", **metadata)

    def log(self, stage: str, seq: int = -1, **extra: object) -> None:
        if self._file is None:
            return
        ts = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
        parts = [ts, "python", f"stage={stage}"]
        if seq >= 0:
            parts.append(f"seq={seq}")
        for key, value in extra.items():
            if value is None:
                continue
            parts.append(f"{key}={value}")
        line = " | ".join(parts) + "\n"
        with self._lock:
            self._file.write(line)
            self._file.flush()

    def close(self) -> None:
        if self._file is None:
            return
        self.log("session_end", seq=-1, extra="source=python")
        with self._lock:
            try:
                self._file.close()
            except OSError:
                pass
        self._file = None

File: architect/Assets/Scripts/PipelineTrace.cs
using System;
using System.Globalization;
using System.IO;
using System.Text;
using UnityEngine;

/// <summary>
/// Append-only text log for pipeline latency (pairs with pose.py --pipeline-trace).
/// Use the same file path on both sides to correlate Python and Unity events by seq + timestamp.
/// </summary>
public static class PipelineTrace
{
    static readonly object Lock = new object();
    static string _path;
    static StreamWriter _writer;

    public static bool Enabled => _writer != null;

    public static void Init(string path)
    {
        Shutdown();
        if (string.IsNullOrWhiteSpace(path))
            return;

        _path = path.Trim();
        try
        {
            var dir = Path.GetDirectoryName(Path.GetFullPath(_path));
            if (!string.IsNullOrEmpty(dir))
                Directory.CreateDirectory(dir);
            _writer = new StreamWriter(_path, append: true, Encoding.UTF8) { AutoFlush = true };
            Log("session_start", -1, "source=unity");
            Debug.Log($"[PipelineTrace] Logging to {_path}");
        }
        catch (Exception e)
        {
            Debug.LogWarning($"[PipelineTrace] Could not open log file '{_path}': {e.Message}");
            _writer = null;
            _path = null;
        }
    }

    public static void Log(string stage, int seq = -1, string extra = null)
    {
        if (_writer == null)
            return;

        var ts = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff", CultureInfo.InvariantCulture);
        var sb = new StringBuilder(128);
        sb.Append(ts).Append(" | unity | stage=").Append(stage);
        if (seq >= 0)
            sb.Append(" | seq=").Append(seq);
        if (!string.IsNullOrEmpty(extra))
            sb.Append(" | ").Append(extra);
        sb.AppendLine();

        lock (Lock)
        {
            try
            {
                _writer.Write(sb.ToString());
            }
            catch (Exception e)
            {
                Debug.LogWarning($"[PipelineTrace] Write failed: {e.Message}");
            }
        }
    }

    public static void Shutdown()
    {
        lock (Lock)
        {
            if (_writer == null)
                return;
            try
            {
                Log("session_end", -1, "source=unity");
                _writer.Dispose();
            }
            catch (Exception)
            {
                // ignore on shutdown
            }
            _writer = null;
            _path = null;
        }
    }
}
